---
title: "DOCUMENTATION"
output:
  html_document: default
  pdf_document: default
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T)
```
## To check:
* check that cohort_name in inventory_patients is actually used

## Welcome
This [Nextflow](https://www.nextflow.io/) pipeline is designed for the de-novo detection of coding and noncoding somatic driver genomic elements in cancer patient cohorts. It currently integrates five advanced calling algorithms: [DIGdriver](https://github.com/maxwellsh/DIGDriver), [dNdScv](https://github.com/im3sanger/dndscv/tree/master), NBR, [MutPanning](https://www.genepattern.org/modules/docs/MutPanning#gsc.tab=0), and [OncodriveFML](https://bbglab.irbbarcelona.org/oncodrivefml/home). DIGdriver, NBR, and OncodriveFML are capable of detecting both coding and noncoding driver genetic elements, whereas dNdScv and MutPanning focus solely on detecting coding drivers. Source code for NBR was provided by Dr. 
[Inigo Martincorena](https://github.com/im3sanger). 

This documentation provides comprehensive instructions on setting up, configuring, and running the pipeline, along with detailed descriptions of the outputs.

## Requirements
* **Nextflow**: The pipeline is written in DSL2 and requires Nextflow version 23.04.2 or higher.
* **Singularity**: All software used in the pipeline is containerized. Interactions with containers are executed via Singularity. The pipeline has been tested with Singularity version 3.8.3.

## Supported genome versions

## Input formats
Ideally, all input files should be in `hg19` coordinates. However, if this is not the case, avoid performing the liftover as it is already implemented in the pipeline. This approach minimizes the potential inconsistencies introduced by the liftover procedure.

### Genomic variants files (mutations)
### Genomic regions of interest
### Mutations multiplicity

### Inventory tables
#### Patients inventory table
The patient inventory table is a comma-separated file that contains detailed information about all participants (patients) in the cohort, including ID, tumor subtype, path to the mutation table, and other relevant data. This table is also used to define specific cohorts of participants for further analysis, i.e. adenocarcinomas, squamous cell carcinomas, pan-cancer, etc. As parallelisation is ensured by the pipeline architecture as well as by Nextflow itself, there is no need to have separate patient inventory tables for each tumor subtype. 

The table below provides an example of a patient inventory table.

| **tumor_subtype** | **participant_id** | **participant_tumor_subtype** | **somatic_genome** | **somatic_path** | **mutmultiplicity_path** | **cn_segments_genome** | **cn_segments_path** | **cohort_name** |
|:-----------------:|:------------------:|:-----------------------------:|:------------------:|:----------------:|:------------------------:|:----------------------:|:--------------------:|:---------------:|
| Adenocarcinoma    | participant_1      | LUAD                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Adenocarcinoma    | participant_2      | LUAD                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Adenocarcinoma    | participant_3      | LUAD                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Squamous_cell    | participant_41      | LUSC                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Squamous_cell    | participant_42      | LUSC                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Squamous_cell    | participant_43      | LUSC                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Panlung    | participant_1      | LUAD                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Panlung    | participant_2      | LUAD                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Panlung    | participant_3      | LUAD                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Panlung    | participant_41      | LUSC                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Panlung    | participant_42      | LUSC                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |
| Panlung    | participant_43      | LUSC                | hg38               | full_path_to_file| full_path_to_file        | hg38                   | full_path_to_file    | GEL            |

where

- **tumor_subtype** *[essential]*: The name of the tumor cohort to be analyzed. For example, all patients with lung adenocarcinomas may be grouped in a cohort named `Adenocarcinoma`. This column must not contain values which are numbers, i.e. "adenocarcinama_1" is allowed value, but "1234" is not. The values in this column must not contain "-" character.
- **participant_id** *[essential]*: The unique ID of a patient, e.g., `participant_1`. Each value of `participant_id` column must be linked to one and only one value of `participant_tumor_subtype` column.
- **participant_tumor_subtype** *[essential]*: The histological subtype of a tumor found in the corresponding participant, e.g., `LUAD` (**Lu**ng **Ad**enocarcinoma). This column must not contain values which are numbers,  i.e. "LUAD_1" is allowed value, but "78" is not.
- **somatic_genome** *[essential]*: The version of the genome in which the coordinates of mutations are specified, e.g., `hg38`. This column must not contain values which are numbers, i.e. "hg38" is allowed value, but "38" is not. The genome version must be the same for all files listed in **somatic_path** column. Ideally, all input files should be in `hg19` coordinates. However, if this is not the case, avoid performing the liftover as it is already implemented in the pipeline. This approach minimizes the potential inconsistencies introduced by the liftover procedure.
- **somatic_path** *[essential]*: The full path on your system (computer/HPC/*etc*) to the file containing **somatic** genetic mutations detected in the tumor of the corresponding individual. For the format of this file, see the section above. Existence of the files in this column will be checked on prior to pipeline execution.
- **mutmultiplicity_path** *[optional]*: The full path on your system (computer/HPC/*etc*) to the file containing mutation multiplicities for the genomic variants defined in the `somatic_path` column. For the format of this file, see the section above. Existence of the files in this column will be checked on prior to pipeline execution (if column is present in the table). 
- **cn_segments_genome** *[optional]*: The version of the genome in which the coordinates of copy number segments are specified, e.g., `hg38`. This column must not contain values which are numbers,  i.e. "hg38" is allowed value, but "38" is not. The genome version must be the same for all files listed in **cn_segments_path** column.
- **cn_segments_path** *[optional]*: The full path on your system (computer/HPC/*etc*) to the file containing copy number segments of the tumor genome found in the corresponding participant. For the format of this file, see the section above. Existence of the files in this column will be checked on prior to pipeline execution (if column is present in the table).
- **cohort_name**  *[essential]*: . This column must not contain values which are numbers, i.e. "GEL_1" is allowed value, but "009" is not.

Cancer cohorts that include multiple histological subtypes (for example, a `pan-lung` cancer cohort may include tumor samples from adenocarcinomas, squamous cell carcinomas, mesotheliomas, neuroendocrine carcinomas, *etc.*) can be defined as shown in lines 7-12 of the table. It is preferable that the **participant_tumor_subtype** column contains the actual histological subtype of the tumor, rather than a "pan-lung" substitute.
 
#### Analysis inventory table
The analysis inventory table is a comma-separated file that defines the genomic regions to be scanned for potential cancer driver elements. It also links tumor subtypes defined in the patient inventory table to these genomic regions of interest. Additionally, the table specifies the software to be used for scanning each genomic region. As parallelisation is ensured by the pipeline architecture as well as by Nextflow itself, there is no need to have separate analysis inventory tables for each tumor subtype. 

The table below provides an example of an analysis inventory table.

| **tumor_subtype** | **software** | **gr_id** | **gr_code** | **gr_file** | **gr_upstr** | **gr_downstr** | **gr_genome** | **gr_excl_id** | **gr_excl_code** | **gr_excl_file** | **gr_excl_upstr** | **gr_excl_downstr** | **gr_excl_genome** | **blacklisted_codes** |

where

- **tumor_subtype** *[essential]*: The name of the tumor cohort to be analyzed. It should match one of the cohort names listed in the `tumor_subtype` column of the [patient inventory table](#Patients-inventory-table). The values in this column must not contain "-" character.
- **software** *[essential]*: The name of the software to be applied to the cohort listed in the `tumor_subtype` column. Permitted values are: `digdriver`, `dndscv`, `mutpanning`, `chasmplus`, `nbr`, and `oncodrivefml`.
- **gr_id** *[essential]*: The name of the genomic region(s) set(s) to be analyzed. This can be any user-defined string, such as "coding" for CDS. The values in this column must not contain "-" character.
- **gr_code** *[essential]*: A code (string) defining the biotypes of component parts of the genomic region(s) set(s) to be analyzed. The biotypes of the component parts are defined via files listed in `gr_file` column.


Permitted values are: `3primeUTR`, `5primeUTR`, `CDS`, `lincRNA`, `lincRNA_promoter`, `lincRNA_ss`, `miRNA`, `misc_RNA`, `promoter`, `rRNA`, `snoRNA`, `snRNA`, and `ss` (splice sites). Genomic regions annotated with the corresponding biological function (e.g., all 3'UTRs in the case of `3primeUTR` or long non-coding RNAs in the case of `lincRNA`) will be extracted from the files specified in `gr_file`, combined gene-wise (collapsed across transcripts in the case of UTRs), and included in the set defined in the `gr_id` column. Each set of genomic regions of interest (defined uniquely by its ID in the `gr_id` column) can comprise one or more `gr_code`s. For example, a `gr_id` named "coding_and_UTRs" can have `CDS`, `3primeUTR`, and `5primeUTR` in the `gr_code` column. Please see below for illustrative examples. For `dNdScv` and `MutPanning` only `CDS` is accepted in this column.
- **gr_file** *[essential]*: The path to a GTF or BED file from which genomic elements of interest should be extracted. The BED file must have the following columns: `chr`, `start`, `end`, `strand`, `gene_id`, `gene_name`, `rCode`, where the `rCode` column contains the biotype of the region as a string, matching one of the values in the `gr_code` column. The GTF file must have the following fields: `gene_name`, `gene_id`, `gene_type`, `gene_biotype`, `transcript_id`, `transcript_type`, `transcript_biotype`. Please refer to the section [genomic regions of interest](#Genomic-regions-of-interest) for more details. Existence of the files listed in this column will be checked prior to pipeline's execution.
- **gr_upstr** *[essential]*: The number of bases upstream of the genomic region of interest to be included. For example, 5'UTRs can be extended upstream by an additional 50bp.
- **gr_downstr** *[essential]*: The number of bases downstream of the genomic region of interest to be included.
- **gr_genome** *[essential]*: The version of the genome in which the genomic coordinates of the regions are presented in the file listed in the `gr_file` column, e.g., `hg38`.
- **blacklisted_codes** *[essential]*: 

In many cases, it is necessary to exclude certain genomic regions from the regions of interest to ensure clarity in the analysis. For example, regions overlapping with the cumulative set of CDS coordinates are typically excluded from splice site regions to prevent contamination of the splice site signal with the signal from CDS. The columns `gr_excl_id`, `gr_excl_code`, `gr_excl_file`, `gr_excl_upstr`, `gr_excl_downstr`, and `gr_excl_genome` define these exclusion regions in the same manner as the columns described above define regions of interest.


    
#' 8) gr_code is acceptable, if gr_file and gr_excl_file are gtf 
#' 9) for promoters and splice sites upstream and downstream are not 0 at the
#'    same time if gtf file is given 
#' 10) within same gr_id, all lines have same value of blacklisted codes
#' 11) within same tumor_subtype, all lines have same value of blacklisted 
#'     codes
#' 12) there are <= 2 genome versions, including target genome version
 
#' 14) gr_id is defined by the same files and upstream/downstream combinations
#'     across tumor_subtypes and software
#' 15) check, that if dndscv is requested, a gtf file is given
#' 16) check, that if digdriver is requested, a gtf file is given, one per 
#'     tumor
#' 17) If union_percentage/intersect_percentage column are given, check that 
#'     it is NA for rows with gr_code equals to CDS, and values in it in range
#'     of 0 to 100 for the other rows 
#  18) if column restrict_to is present in the inventory check that its values
#      are either empty or known. 
#  19) if column restrict_to is present in the inventory check that there is 
#      only one value per every tumor_subtype

#### Black and white listed genomic regions inventory table
### Other files

## Outputs
## Notes

do not forget to add the check that gtf does not contain "--" and that there is either protein_id or ccdsid
Add GTEx and TCGA inventories too, but only for the postprocessing script

Note on NBR: it will give no results on TCGA due to no overlap of mutations and
neutral regions. It is truly for WGS only.
Note on CHASMplus: to run proper model, change containers in config

https://github.com/OpenGenomics/mc3/blob/aff72c9696031c2f2bdedbb549ceb8102e0c0ef0/refdocs_tree.txt#L17
https://data.broadinstitute.org/snowman/hg19/annotation/
wget https://data.broadinstitute.org/snowman/hg19/Homo_sapiens_assembly19.fasta
wget https://data.broadinstitute.org/snowman/hg19/annotation/Homo_sapiens_assembly19.gtf

because DIGdriver can only handle **reference genome in UCSC format**, we have to add 'chr' to the name
of each chromosome in fasta file. no need to do that for your mutations or gtf regions, the code
will take care of this.

How cgc_intogen_mc3_knownCancerGenes and olfactory_barnes_2020.csv was created? Do not forget that both gene_id and gene_name is used to match known genes to the results of analysis

## Oncodrivefml
downloading bgdata is done properly if and only if example is run.
is big 17GB

https://oncodrivefml.readthedocs.io/en/latest/includes.html#readme-example

## DIGdriver
mkdir -p $DIG_MODELS_DIR
cd $DIG_MODELS_DIR

# since durng execution DIGdriver changes content of its models and element 
# files, files will be made read only and in the run script copied for every 
# execution and run.

wget http://cb.csail.mit.edu/cb/DIG/downloads/mutation_maps/Lung_tumors_SNV_MNV_INDEL.Pretrained.h5
mv Lung_tumors_SNV_MNV_INDEL.Pretrained.h5 digdriver-PANCAN-SNV_MNV_INDEL.Pretrained.h5
chmod 0444 digdriver-PANCAN-SNV_MNV_INDEL.Pretrained.h5

wget http://cb.csail.mit.edu/cb/DIG/downloads/mutation_maps/Lung-AdenoCA_SNV_MNV_INDEL.Pretrained.h5
mv Lung-AdenoCA_SNV_MNV_INDEL.Pretrained.h5 digdriver-ADENOCARCINOMA-SNV_MNV_INDEL.Pretrained.h5
chmod 0444 digdriver-ADENOCARCINOMA-SNV_MNV_INDEL.Pretrained.h5

wget http://cb.csail.mit.edu/cb/DIG/downloads/mutation_maps/Lung-SCC_SNV_MNV_INDEL.Pretrained.h5
mv Lung-SCC_SNV_MNV_INDEL.Pretrained.h5 digdriver-SQUAMOUS_CELL-SNV_MNV_INDEL.Pretrained.h5
chmod 0444 digdriver-ADENOCARCINOMA-SNV_MNV_INDEL.Pretrained.h5

wget http://cb.csail.mit.edu/cb/DIG/downloads//dig_data_files/element_data.h5
mv element_data.h5 digdriver-element_data.h5
chmod 0444 digdriver-element_data.h5

cd ../

## CHASMplus
You have to re-create container with your own models. Otherwise it gets too
big. Don't forget to replace it in conf file.

## MISC
Multiple expression tables are supported

Do not forget to add chmod +x on R file into tutorial
note on binding for singularity containers
Grep Use.For.Plots because it can still appear in scripts
Add a small workflow on how to prepare ensemble gtf to the biotyped gtf
Provide with the pipeline biotype annotated GTF files on hg38 and on hg19
REMEMBER THAT:
Nextflow implementation does not filter files by columns 'Use.For.Plots' and 'Use.For.Plots.Indelâ€™

## Organising your data
* inventory folder
* _assets folder
* additional_db folder
* results folder
target_genome_path

## List of all code words & parameters
cancer_subtype
min_n_participants

PANCAN
min_depth
min_tumor_vac
max_germline_vaf
max_germline_vac
max_n_vars
target_genome_version
zip
cores - maybe various options, because requirements may vary

## Inventory

We do not recommend running both dNdScv and NBR on CDS

### Participants inventory

Essential file. Recommended name: . It is recommended, not necessary.

| **participant_id** | **tumor_subtype** |   **somatic_path**   | **somatic_genome** | **cohort_name** |
|:------------------:|:-----------------:|:--------------------:|:------------------:|:---------------:|
| TCGA-05-4244       | LUAD              |  *full path to file* |      hg19          |      TCGA       |
| TCGA-05-4249       | LUAD              |  *full path to file* |      hg19          |      TCGA       |
| TCGA-05-4250       | LUAD              |  *full path to file* |      hg19          |      TCGA       |
| TCGA-18-3406       | LUSC              |  *full path to file* |      hg19          |      TCGA       |
| TCGA-18-3407       | LUSC              |  *full path to file* |      hg19          |      TCGA       |

where
    
- **participant_id**: unique ID of a patient, i.e. `TCGA-05-4244`. It should
                      appear once and only once in the table.
- **tumor_subtype**: tumor subtype of that patient, i.e. `LUAD`. 
                     tumor_subtype and tumor_type columns 
                     **should not contain "-" sign** as it is used within the 
                     analysis for other purposes. For example, NEURO-ENDOCRINE
                     in tumor_type column will be flagged, and NEURO_ENDOCRINE
                     is allowed. Also, values in tumor_subtype column can not
                     be numbers.
- **somatic_path**: a full path on your system (computer/server) to the 
                    file with **somatic** genetic mutations detected in **that
                    individual**. For the format of that file, see below.
- **somatic_genome**: a genome version to which data from that patient were
                      mapped, i.e. `hg19`.
- **cohort_name** : name of the cohort
                      
**At least 9 patients per tumor type is recommended.** Although this is a bare
minimum and does not guarantee best results. All other columns in the table 
will be ignored. **Comma separated**

### Analysis inventory

| **tumor_subtype** | **restrictedTest** | **software** | **gr_id** | **gr_code** | **gr_file** | **gr_upstr** | **gr_downstr** | **gr_genome** | **gr_base** | **gr_excl_id** | **gr_excl_code** | **gr_excl_file** | **gr_excl_upstr** | **gr_excl_downstr** | **gr_excl_genome** | **gr_excl_base** | **blacklisted_codes** | **union_percentage** | **intersect_percentage** |
|:-----------------:|:------------------:|:------------:|:---------:|:-----------:|:---------------------------------------------------------------------------------------------------------------------------:|:------------:|:--------------:|:-------------:|:-----------:|:--------------:|:----------------:|:---------------------------------------------------------------------------------------------------------------------------:|:-----------------:|:-------------------:|:------------------:|:----------------:|:---------------------:|:--------------------:|:------------------------:|
| LUAD              | FALSE              | chasmplus    | CDS       | CDS         | /Users/maria/Desktop/BitBucket//Noncoding_drivers/_assets/additional_db_processed_2023-02-21/hg38.refGene_biotyped_hg38.gtf | 0            | 0              | hg19          | 0           | NA             | NA               | NA                                                                                                                          | NA                | NA                  | NA                 | NA               | CRG;DAC;DUKE          | NA                   | NA                       |
| LUAD              | FALSE              | dndscv       | CDS       | CDS         | /Users/maria/Desktop/BitBucket//Noncoding_drivers/_assets/additional_db_processed_2023-02-21/hg38.refGene_biotyped_hg38.gtf | 0            | 0              | hg19          | 0           | NA             | NA               | NA                                                                                                                          | NA                | NA                  | NA                 | NA               | CRG;DAC;DUKE          | NA                   | NA                       |
| LUAD              | FALSE              | mutpanning   | CDS       | CDS         | /Users/maria/Desktop/BitBucket//Noncoding_drivers/_assets/additional_db_processed_2023-02-21/hg38.refGene_biotyped_hg38.gtf | 0            | 0              | hg19          | 0           | NA             | NA               | NA                                                                                                                          | NA                | NA                  | NA                 | NA               | CRG;DAC;DUKE          | NA                   | NA                       |
| LUAD              | FALSE              | oncodrivefml | CDS       | CDS         | /Users/maria/Desktop/BitBucket//Noncoding_drivers/_assets/additional_db_processed_2023-02-21/hg38.refGene_biotyped_hg38.gtf | 0            | 0              | hg19          | 0           | NA             | NA               | NA                                                                                                                          | NA                | NA                  | NA                 | NA               | CRG;DAC;DUKE          | NA                   | NA                       |
| LUAD              | FALSE              | oncodrivefml | ss        | ss          | /Users/maria/Desktop/BitBucket//Noncoding_drivers/_assets/additional_db_processed_2023-02-21/hg38.refGene_biotyped_hg38.gtf | 20           | 6              | hg19          | 0           | CDS            | CDS              | /Users/maria/Desktop/BitBucket//Noncoding_drivers/_assets/additional_db_processed_2023-02-21/hg38.refGene_biotyped_hg38.gtf | 0                 | 0                   | hg19               | 0                | CRG;DAC;DUKE          | 50                   | 80                       |

where

- **tumor_subtype**: 
- **restrictedTest**: optional
- **software**: 
- **gr_id**: 
- **gr_code**: 
- **gr_file**: 
- **gr_upstr**: 
- **gr_downstr**: 
- **gr_genome**: 
- **gr_excl_id**: 
- **gr_excl_code**: 
- **gr_excl_file**: 
- **gr_excl_upstr**: 
- **gr_excl_downstr**: 
- **gr_excl_genome**: 
- **blacklisted_codes**: 
- **union_percentage**: optional
- **intersect_percentage**: optional

**Comma separated**

### Black/white list inventory
|**list_name**|**file_path** |**file_genome**| **file_type**  |**score_column**|**min_value**| **max_value** |
|:-----------:|:------------:|:-------------:|:--------------:|:--------------:|:-----------:|:-------------:|
|             |              |               |                |                |             |               |

where

- **file_path**: 
- **file_type**: 
- **score_column**: 
- **min_value**: 
- **max_value**: 
- **file_genome**: 

1) files from the table exist 2) codes in file_type are white or black 
3) min_value and max_value are given is score_column is given 4) genomes versions of 
files are allowed ones.

We highly recommend to bring this files to target genome version as this process
can be long and high memory

**Comma separated**

## Input files

### Somatic mutations table(s)
In case of annovar format, it's not possible to filter based on VAF/VAC, etc.

#### Annovar-like:

|key|chr|start|stop|ref|var|Gene.refGene|Func.refGene|ExonicFunc.refGene|GeneDetail.refGene|AAChange.refGene|Use.For.Plots|Use.For.Plots.Indel|t_depth|t_ref_count|t_alt_count|n_depth|n_ref_count|n_alt_count|t_maxVAF|
|---|---|-----|----|---|---|------------|------------|------------------|------------------|----------------|-------------|-------------------|-------|-----------|-----------|-------|-----------|-----------|--------|
|trololo|   |     |    |   |   |            |            |                  |                  |                |             |                   |       |           |           |       |           |           |        |


after t_depth columns are optional

#### MAF:

|Tumor_Sample_Barcode|Chromosome|Start_Position|End_Position|Reference_Allele|Tumor_Seq_Allele2|Gene|Variant_Classification|Amino_acids|t_depth|t_ref_count|t_alt_count|n_depth|n_ref_count|n_alt_count|t_maxVAF|
|:------------------:|:--------:|:------------:|:----------:|:--------------:|:---------------:|:--:|:--------------------:|:---------:|:-----:|:---------:|:---------:|:-----:|:---------:|:---------:|:------:|
|      trololo       |          |              |            |                |                 |    |                      |           |       |           |           |       |           |           |        |


## Configuring the pipeline

## A note on genome versions
Mutpanning runs on hg19 only. Liftover shatters CN regions into lots of small 
ones which can lead to overestimation of number of driver mutations. Therefore
it is best to perform match between genetic elements and CN on the same genome
version CN were done.

#We have to store somewhere IDs of patients which were filtered out as hypermutated

GTF FILE SHOULD NOT CONTAIN GENE NAMES/IDS WITH -- IN THEM. FOR CURRENT HG37
I MANUALLY REPLACED THEM WITH -
